spec/models/adviser_spec.rb

```diff
@@ -163,4 +163,75 @@ def queue_contains_a_job_for(job_class)
       end
     end
   end
+
+  describe 'after_save :check_for_changes' do
+    let(:original_firm) { create(:firm) }
+    let(:receiving_firm) { create(:firm) }
+    subject { create(:adviser, firm: original_firm) }
+
+    def save_with_commit_callback(model)
+      model.save!
+      model.run_callbacks(:commit)
+    end
+
+    context 'when the firm has changed' do
+      it 'triggers reindexing of the adviser and new firm' do
+        expect(GeocodeAdviserJob).to receive(:perform_later).with(subject)
+        subject.firm = receiving_firm
+        save_with_commit_callback(subject)
+      end
+
+      it 'triggers reindexing of the original firm' do
+        expect(IndexFirmJob).to receive(:perform_later).once().with(original_firm.id)
+        subject.firm = receiving_firm
+        save_with_commit_callback(subject)
+        subject.run_callbacks(:commit)
+      end
+    end
+  end
+
+  describe '.move_to' do
+    let(:original_firm) { create(:firm_with_advisers) }
+    let(:receiving_firm) { create(:firm_with_advisers) }
+
+    it 'moves a batch of advisers to another firm' do
+      advisers_to_move = original_firm.advisers.limit(2)
+      advisers_to_move.move_to(receiving_firm)
+
+      expect(advisers_to_move[0].firm).to be(receiving_firm)
+      expect(advisers_to_move[1].firm).to be(receiving_firm)
+
+      receiving_firm.reload
+      original_firm.reload
+
+      expect(original_firm.advisers.count).to be(1)
+      expect(receiving_firm.advisers.count).to be(5)
+      expect(receiving_firm.adviser_ids).to include(advisers_to_move[0].id,
+                                                    advisers_to_move[1].id)
+      expect(original_firm.adviser_ids).not_to include(advisers_to_move[0].id,
+                                                       advisers_to_move[1].id)
+    end
+
+    context 'when one of the move operations fails' do
+      let(:advisers_to_move) { original_firm.advisers.limit(3) }
+      let(:invalid_record_index) { 1 }
```

Comment:
Ok considered this, but I've decided I want it to be really obvious that I'm invalidating a record in the set I'm about to move.